### PR TITLE
Global code refactor

### PR DESCRIPTION
**Pull request by torbacz**
## Motivation and Context

- Partially resolves nanoFramework/Home#1130

## How Has This Been Tested?

- Local tests from POC project.
## Screenshots
image

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by Ellerbach** [comment]
Thanks for the refactoring. I bet you are intensively testing all the changes, correct? :-)

**Comment by torbacz** [comment]
@Ellerbach 
This one, not yet. Can't run tests project locally. That's why it's draft :)

**Comment by josesimoes** [comment]
@torbacz why is that? Any error message or behavior that you're seeing to help understand this?

File: source/TestAdapter/Discover.cs [hunk 1]
```diff
@@ -120,38 +120,40 @@ public static List<TestCase> FindTestCases(string source)
             Type[] allTypes = test.GetTypes();
             foreach (var type in allTypes)
             {
-                if (type.IsClass)
+                if (!type.IsClass)
```
Comment (josesimoes) [inline source/TestAdapter/Discover.cs:123]: I get it that this is more a "code style preference" but I always find it hard in my mind to follow code when making negative comparisons... 😅

File: source/TestAdapter/TestObjectHelper.cs [hunk 2]
```diff
@@ -28,11 +28,9 @@ public static string GetNanoClrLocation()
             {
                 return nanoClrFullPath;
             }
-            else
-            {
-                var inititialDir = new DirectoryInfo(Path.GetDirectoryName(thisAssemblyDir));
-                return FindNanoClr(inititialDir);
-            }
+
```
Comment (josesimoes) [inline source/TestAdapter/TestObjectHelper.cs:31]: Your change will have the method perform the same way, no questions about it! 😃 But (again!) I prefer to have these coded with the `else`. It greatly improves code readability as it shows the intent of the `else` block being executed as an alternative to the `if` clause. 
When you remove the `else`, you're "closing" the matter and "moving along", which makes that code block "loose" from the `if` from which it belongs to.

File: source/TestAdapter/TestObjectHelper.cs [hunk 3]
```diff
@@ -45,13 +43,10 @@ private static string FindNanoClr(DirectoryInfo initialPath)
...
                 }
-                else
-                {
-                    return FindNanoClr(dir);
-                }
+                return FindNanoClr(dir);
             }
 
-            return string.Empty;
+            throw new FileNotFoundException($"Unabl to find nanoClr.");
```
Comment (josesimoes) [inline source/TestAdapter/TestObjectHelper.cs:49]:
```suggestion
            throw new FileNotFoundException($"Unable to find nanoCLR.");
```

File: source/TestFramework/TestFramework.cs [hunk 4]
```diff
@@ -22,10 +22,8 @@ public static void SkipTest(string message = "")
             {
                 throw new SkipTestException();
             }
-            else
-            {
-                throw new SkipTestException(message);
-            }
+
+            throw new SkipTestException(message);
```
Comment (josesimoes) [inline source/TestFramework/TestFramework.cs:26]: Same here...

**Comment by torbacz** [comment]
First standard issues with local tests - fixed by adjusting nfproj and nano.runsettings. 
image
image

Now
image

**Comment by torbacz** [comment]
@josesimoes 
I know it may be a confusing at first but it's common refactoring technique to keep code clean. For simple conditions like here it may not make sens right now, but if you add few more conditions in future it easier to read that way. Check this article: https://betterprogramming.pub/refactoring-guard-clauses-2ceeaa1a9da (you can skip text, just check code snippets) 
Also, when there is less code the reader is happier 😄 

But If you want, I can revert this changes. Up to you.

File: source/TestAdapter/Discover.cs (hunk 1 above)
Comment (torbacz) [reply source/TestAdapter/Discover.cs:123]: Is it better now?

**Comment by torbacz** [comment]
And the "POC" tests results
image

**Comment by Ellerbach** [comment]
Can you add a test in the POC with the data row? Not sure there is one. And show how it is displayed then in the screen capture. Just want to make sure that one will work as well.

On the refactoring and reverse conditions, it's a matter of habits. I'm not the biggest fan neither but if this is a bit more simple and still work, I'm ok with it.

**Comment by torbacz** [comment]
@Ellerbach 
Already there, check the results from previous screen. They are displayed as "MethodName(index[0-x])"
Again, can't run tests locally... "No suitable test runner found"

**Comment by josesimoes** [comment]
> @josesimoes I know it may be a confusing at first but it's common refactoring technique to keep code clean. For simple conditions like here it may not make sens right now, but if you add few more conditions in future it easier to read that way. Check this article: https://betterprogramming.pub/refactoring-guard-clauses-2ceeaa1a9da (you can skip text, just check code snippets) Also, when there is less code the reader is happier 😄

Thanks for sharing. I understand some of Carlos points there, but I don't necessarily agree with all of them. 😁 

In particular in this PR: I'm OK with your changes applying to those large code refactors on which you switch the IF clause and return immediately, then the large code block comes next. A totally different thing is applying that to those small code blocks. On those I really think that keeping the "positive" IF and ELSE do improve code readability. I admit this is very much related with code style and personal preference, so let's ask for another opinion and move forward with this. 😉  
@Ellerbach care to weigh in?

**Comment by Ellerbach** [comment]
@josesimoes you have a code style preference, I have a code style preference (close to yours), @torbacz have a code style preference. Everyone has. And we don't have specific rules for this one about checking conditions and imbricated if. We do have others commonly admitted with explicit bracelet, naming conventions, documentation, etc.
I am fine with this rework of the code because I still can read, yes, I have a make a little bit more efforts because of the reverse checks but that's ok. And there are optimizations we wanted (like typeof checks for the types, etc).
And the code is clean, readable.
So fine with this!

**Review by Ellerbach — APPROVED** [review]
Thanks for the adjustments even if code style is a bit different than mine :-)

**Review by josesimoes — APPROVED** [review]
LGTM

**Pull request merged**